### PR TITLE
#7284: added mamba perf reports scripts to CI

### PR DESCRIPTION
**Pull request by kpaigwar**
* Moved model from experimental to demos
* Added device and e2e performance tests
* Cleaned up redundant files
* Update README

### CODE REVIEW

File: tests/scripts/run_models.sh [hunk 1]
```diff
@@ -18,10 +18,8 @@ if [[ $ARCH_NAME == "wormhole" ]]; then
 fi
 
 if [[ $ARCH_NAME == "wormhole_b0" ]]; then
-  env pytest models/experimental/mamba/tests/test_full_model.py
-  env pytest models/experimental/mamba/tests_opt/test_full_model.py
-  env pytest models/experimental/mamba/tests/test_benchmarks.py
-  env pytest models/experimental/mamba/tests/test_demo.py
+  env pytest models/demos/mamba/tests/test_full_model.py::test_mamba_inference_ci
```
Comment (esmalTT) [inline tests/scripts/run_models.sh:21]: I think we should run everything in our test folder except for perf. I think you could clean up those extra functions in `test_full_model.py` and then just list all of the test files we have (minus perf) here. 

What might be even better - rename `test_mamba_perf` to `run_mamba_perf` and we can just list `pytest models/demos/mamba/tests/test_*` here. It would be nice to avoid listing these things explicitly so when we add new test files they go into CI automatically.

**Review by esmalTT — CHANGES_REQUESTED** [review]

File: models/demos/mamba/demo/demo.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    if args.prompt is not None:
+        prompts = [args.prompt] * num_users
+    else:
+        with open(args.prompts_file, "r") as f:
+            prompts = json.load(f)
+
+    run_demo(prompts, args.device, generated_sequence_length=args.gen_len)
+
+
+if __name__ == "__main__":
```
Comment (aguhaTT) [inline models/demos/mamba/demo/demo.py:151]: we should use pytest.

File: models/demos/mamba/demo/demo.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    use_cache: bool = True,
+):
+    batch_size = len(prompts)
+    assert batch_size == 32, "32 prompts are required"
+    device = None
+
+    logger.info(f"Running Mamba demo (weights='{model_version}') on '{model_type.upper()}' with batch={batch_size}")
+
+    if model_type == "cpu":
+        model = get_cpu_reference_model(model_version, batch_size)
```
Comment (aguhaTT) [inline models/demos/mamba/demo/demo.py:88]: not sure how to handle this if using pytest.

Comment (aguhaTT) [reply models/demos/mamba/demo/demo.py:88]: can you check with other models?

File: models/demos/mamba/demo/demo.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        model = get_tt_metal_model(model_version, device, use_cache, batch_size)
+    else:
+        raise RuntimeError("Invalid model type was encountered")
+
+    tokenizer = get_tokenizer()
+
+    sequences: torch.Tensor = tokenizer(prompts, return_tensors="pt", padding=True).input_ids
+
+    # prefill
+    prefill_iterations = sequences.shape[1] - 1
```
Comment (aguhaTT) [inline models/demos/mamba/demo/demo.py:101]: I am wondering if we should prefill on CPU and push the resulting hidden state to silicon. This will make the demo faster and help focus on what is important.

File: models/demos/mamba/demo/demo.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    return MambaDecode.from_pretrained(version, batch_size=batch_size)
+
+
+def get_tt_metal_model(version: MambaPretrainedModelName, device: ttnn.Device, use_cache: bool, batch_size: int):
+    from models.demos.mamba.tt.full_model import MambaTT
+    from models.demos.mamba.tt import model_config
+
+    reference_model = get_cpu_reference_model(version, batch_size=batch_size)
+    if use_cache:
+        cache_path = f"/tmp/{version}"
```
Comment (aguhaTT) [inline models/demos/mamba/demo/demo.py:32]: is this the correct path for demo version? what are other models doing?

**Review by aguhaTT — COMMENTED** [review]
I have done a partial review. Full review coming soon. Great job, by the way.

**Review by eyonland — APPROVED** [review]

File: models/demos/mamba/demo/demo.py (hunk 2 above)
Comment (kpaigwar) [reply models/demos/mamba/demo/demo.py]: Updated to pytest

File: models/demos/mamba/demo/demo.py (hunk 3 above)
Comment (kpaigwar) [reply models/demos/mamba/demo/demo.py]: Yeah, CPU case isnot handled by other models. Switched to WH only mode

File: models/demos/mamba/demo/demo.py (hunk 4 above)
Comment (kpaigwar) [reply models/demos/mamba/demo/demo.py:103]: Pushing this work to next PR

File: models/demos/mamba/demo/demo.py (hunk 5 above)
Comment (kpaigwar) [reply models/demos/mamba/demo/demo.py]: Others models are setting path in model args. Set default to there model directory. Added a method to parameterize cache_dir.

File: models/demos/mamba/demo/demo.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    parser = argparse.ArgumentParser(description="Run inference benchmarks on set of supported models")
+    parser.add_argument("--device", choices=["cpu", "wh"], default="wh", help="The device under test")
+    parser.add_argument("--prompts_file", default="prompts.json", help="prompts file to run the demo on")
+    parser.add_argument("--prompt", type=str, default=None, help="single user prompt to run the demo on")
+    parser.add_argument("--gen_len", type=int, default=80, help="length of generated sequence")
+    # Example:
+    # python demo.py --device wh --prompt "Hello, how are you?"
+
+    args = parser.parse_args()
+    num_users = 32
```
[the remark this answers is not in the document]
Comment (kpaigwar) [reply models/demos/mamba/demo/demo.py]: Accounted

File: models/demos/mamba/demo/demo.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    if args.prompt is not None:
+        prompts = [args.prompt] * num_users
+    else:
+        with open(args.prompts_file, "r") as f:
+            prompts = json.load(f)
```
[the remark this answers is not in the document]
Comment (kpaigwar) [reply models/demos/mamba/demo/demo.py]: run_mamba_demo has assert for 32 prompts. Provided instructions in README for custom prompt file.

File: tests/scripts/run_models.sh (hunk 1 above)
Comment (kpaigwar) [reply tests/scripts/run_models.sh]: Added more unit tests in nightly.

File: models/demos/mamba/tests/test_full_model.py [hunk 8]
```diff
@@ -88,3 +103,35 @@ def test_mamba_model_inference(
...
         assert does_pass, f"PCC value is lower than {pcc}"
+
+
+@skip_for_grayskull("Not supported on Grayskull")
+def test_mamba_inference_ci(
+    device: ttnn.Device,
+    use_program_cache,
+    model_version="state-spaces/mamba-2.8b",
+    batch=32,
+    pcc=0.92,
```
[the remark this answers is not in the document]
Comment (kpaigwar) [reply models/demos/mamba/tests/test_full_model.py:108]: This test name has been changed to test_inference_loop. Its job is to test correctness of logits after 10th iterations and also check if it had any hangs or OOM

File: models/demos/mamba/tests/test_full_model.py [hunk 9]
```diff
@@ -88,3 +103,35 @@ def test_mamba_model_inference(
     if not does_pass:
         logger.warning("Mamba output failed")
         assert does_pass, f"PCC value is lower than {pcc}"
+
+
+@skip_for_grayskull("Not supported on Grayskull")
+def test_mamba_inference_ci(
```
[the remark this answers is not in the document]
Comment (kpaigwar) [reply models/demos/mamba/tests/test_full_model.py]: Simplified the tests suite in full_model.py. Now it has mainly three tests 
1. test_inference (iterations=1, layers=64)
2. test_inference_loop (iterations=10, layers=64) --------> called by fast dispatch nightly ci
3. test_device_perf (iterations: [1, 2], layers=1) ------> called by device_perf_ci with an option for warmup to collect tracy host as well.
Note : For pytest to collect tests the function starts with "test_"

**Comment by kpaigwar** [comment]
Mamba Device Perf CI Passed ✅
https://github.com/tenstorrent/tt-metal/actions/runs/8715874745/job/23908455156#step:7:359
All Mamba Nightly Tests Passed ✅ 
https://github.com/tenstorrent/tt-metal/actions/runs/8715860917/job/23908496693#step:9:173

**Review by aguhaTT — APPROVED** [review]

**Review by esmalTT — APPROVED** [review]

File: models/demos/mamba/demo/demo.py [hunk 10]
```diff
@@ -0,0 +1,136 @@
+# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
```
[the remark this answers is not in the document]
Comment (kpaigwar) [reply models/demos/mamba/demo/demo.py:1]: e2e perf test is a copy of our demo with display tokens turned off. So it should protect our demo.

[the remark this answers is not in the document]
Comment (esmalTT) [reply models/demos/mamba/demo/demo.py:1]: Its copy pasted though isn't it?

[the remark this answers is not in the document]
Comment (esmalTT) [reply models/demos/mamba/demo/demo.py:1]: We need to actually invoke the `run_mamba_demo` function.

File: tests/scripts/run_models.sh [hunk 11]
```diff
@@ -18,10 +18,11 @@ if [[ $ARCH_NAME == "wormhole" ]]; then
...
 if [[ $ARCH_NAME == "wormhole_b0" ]]; then
-  env pytest models/experimental/mamba/tests/test_full_model.py
-  env pytest models/experimental/mamba/tests_opt/test_full_model.py
-  env pytest models/experimental/mamba/tests/test_benchmarks.py
-  env pytest models/experimental/mamba/tests/test_demo.py
+  env pytest models/demos/mamba/tests/test_mamba_ssm.py
+  env pytest models/demos/mamba/tests/test_mamba_block.py
+  env pytest models/demos/mamba/tests/test_residual_block.py
+  env pytest models/demos/mamba/tests/test_full_model.py::test_inference_loop
+  env pytest models/demos/mamba/tests/test_benchmarks.py
```
[the remark this answers is not in the document]
Comment (kpaigwar) [reply tests/scripts/run_models.sh:25]: Good point, will add them too

File: models/demos/mamba/demo/demo.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (kpaigwar) [reply models/demos/mamba/demo/demo.py:1]: Yeah correct, in that case we need to add profiling mode in run_mamba_demo.

**Comment by kpaigwar** [comment]
@tapspatel and @tt-rkim could you please review if its ready to merge?

File: tests/scripts/run_performance.sh [hunk 12]
```diff
@@ -61,6 +61,8 @@ run_perf_models_llm_javelin() {
 
     if [ "$tt_arch" == "wormhole_b0" ]; then
         env pytest models/demos/mistral7b/tests -m $test_marker
+
+        env pytest models/demos/mamba/tests -m $test_marker
```
Comment (tt-rkim) [inline tests/scripts/run_performance.sh:65]: Can you please put this command above mistral7b and post a model perf e2e regressions pipeline link where this is passing?

File: tests/scripts/run_models.sh [hunk 13]
```diff
@@ -18,10 +18,14 @@ if [[ $ARCH_NAME == "wormhole" ]]; then
...
 
 if [[ $ARCH_NAME == "wormhole_b0" ]]; then
-  env pytest models/experimental/mamba/tests/test_full_model.py
-  env pytest models/experimental/mamba/tests_opt/test_full_model.py
-  env pytest models/experimental/mamba/tests/test_benchmarks.py
-  env pytest models/experimental/mamba/tests/test_demo.py
+  env pytest models/demos/mamba/tests/test_mamba_ssm.py
+  env pytest models/demos/mamba/tests/test_mamba_block.py
+  env pytest models/demos/mamba/tests/test_residual_block.py
+  env pytest models/demos/mamba/tests/test_full_model.py::test_inference_loop
```
Comment (tt-rkim) [inline tests/scripts/run_models.sh:24]: Can you please target only a filename? If there are multiple tests in this file, please move them to a separate file

Comment (kpaigwar) [reply tests/scripts/run_models.sh]: Refractored and added a separate test file

**Comment by kpaigwar** [comment]
@tt-rkim The three pipelines are running will post the results when they get completed

**Comment by tt-rkim** [comment]
sounds great thank you

**Comment by kpaigwar** [comment]
> sounds great thank you

Oh need to re run all the pipelines with rebase as pytest didn't execute any tests

**Comment by kpaigwar** [comment]
Nightly Passed ✅ https://github.com/tenstorrent/tt-metal/actions/runs/8744459895/job/23998087377#step:9:316
Model Perf ✅ https://github.com/tenstorrent/tt-metal/actions/runs/8744462878/job/23997384470#step:7:2260
Device Perf ❌ failed on SD before mamba
Although it had passed before ✅ https://github.com/tenstorrent/tt-metal/actions/runs/8715874745/job/23908455156#step:7:329
@tt-rkim I think mamba is ready to merge.

**Review by tt-rkim — APPROVED** [review]
